### PR TITLE
Fetch closest sensor

### PR DESCRIPTION
**Pull request by lickel**
Modify the existing geolocation logic to work with the new PurpleAir API.

### CODE REVIEW

File: purpleair-aqi.js [hunk 1]
```diff
@@ -134,38 +134,32 @@ async function getSensorId() {
     const seLat = latitude - BOUND_OFFSET;
     const nwLng = longitude - BOUND_OFFSET;
     const seLng = longitude + BOUND_OFFSET;
-    const req = new Request(
-      `${API_URL}/v1/sensors?fields=name,latitude,longitude,location_type&max_age=3600&location_type=0&nwlat=${nwLat}&selat=${seLat}&nwlng=${nwLng}&selng=${seLng}`
```
Comment (lickel) [inline purpleair-aqi.js:138]: Don't need to send the location_type since you're filtering by it anyways.

File: purpleair-aqi.js [hunk 2]
```diff
@@ -134,38 +134,32 @@ async function getSensorId() {
...
       );
       if (distanceFromLocation < closestDistance) {
         closestDistance = distanceFromLocation;
         closestSensor = location;
       }
     }
 
-    const id = closestSensor[sensorIdField];
-    cacheData("sensor.json", { id, updatedAt: Date.now() });
+    const [id, name, lat, long] = closestSensor;
```
Comment (lickel) [inline purpleair-aqi.js:159]: This "destructures" the array into the 4 fields, which I then use to to create an easier to use object.

File: purpleair-aqi.js [hunk 3]
```diff
@@ -206,24 +200,23 @@ function haversine(start, end) {
...
+  req.headers = {"X-API-Key": API_KEY} ;
 
   let json = await req.loadJSON();
 
   try {
     // Check that our results are what we expect
     if (json && json.sensor) {
       console.log(`Sensor data looks good, will cache.`);
-      const sensorData = { json, last_seen: Date.now() }
+      const sensorData = { json, updatedAt: Date.now() }
```
Comment (lickel) [inline purpleair-aqi.js:213]: If you want to rename this to `last_seen` we'd need to change all of the `updatedAt`s, as this is just the local cache timestamp name.

File: purpleair-aqi.js [hunk 4]
```diff
@@ -459,11 +452,13 @@ function calculateLevel(aqi) {
...
  * Get the AQI trend
  *
- * @param {{ v1: number; v3: number; }} stats
+ * @param {{ pm2.5: number; pm2.5_10minute: number; }} stats
  * @returns {string}
  */
-function getAQITrend({ 'pm2.5': partLive, 'pm2.5_10minute': partTime }) {
-    const partDelta = partTime - partLive;
+function getAQITrend(stats) {
+  const partLive = stats["pm2.5"];
```
Comment (lickel) [inline purpleair-aqi.js:459]: TextMate was choking on the magic destructuring stuff, so I just moved it into the function.

File: purpleair-aqi.js [hunk 5]
```diff
@@ -486,8 +481,12 @@ async function run() {
   const listWidget = new ListWidget();
   listWidget.useDefaultPadding();
 
-try{
-  const sensorId = await getSensorId();
+  try {
+    if (!API_KEY) {
+      throw `You need a PurpleAir API Key for this widget.`;
```
Comment (lickel) [inline purpleair-aqi.js:486]: This was all that we need to show that new error state.
It's red text on a white background.

**Comment by johndoeaki** [comment]
- [ ] approved

**Pull request merged**